ratik/src/main/java/org/opencadc/ratik/ArtifactValidator.java

```diff
@@ -168,7 +166,31 @@ protected void validateLocal(Artifact local)
         Artifact remote = getRemoteArtifact(local.getURI());
         if (remote != null) {
             log.info(String.format("delete: %s %s reason: local filter policy change", local.getID(), local.getURI()));
-            //TODO do action
+            try {
```

Comment:
Only one review comment on this particular block: ArtifactValidator lines 168-195. Please concentrate on fixing this one block before making similar changes in others I haven't looked at.

In general if you using an explicit transaction, you must have a try/catch/finally. Almost always: the startTransaction should be the first thing in the try and the commitTransaction should be the last thing in the try. There must be a catch(Exception) that does a rollback after any error, and there must be a finally that check if the transaction is open and rollback and report error very loudly. See minoc PutAction (near the bottom).

The log.info for this case (currently line 168) should be the line right before `artifactDAO.delete(...)` so it only gets printed if something actually happens.

The EntityNotFoundException from the lock call means that sometime between reading the Artifact from the database and now someone else deleted it. Since you only need the entity ID you do not need `get(local.getID())` again and (after the lock) local can never be null, so just proceed to do the actions and commit the transaction.

In the catch block, you may have to be more careful in calling rollback(). Some code re-org above that puts a new failure mode inside the try but before the start (or after the commit) will cause the rollback to fail and be really confusing. For example, in minoc PutAction: there is some code in the try block after the commit (legit in that case) that could throw, and then the rollback would fail. that code was moved and the catch block not updated to deal with a failure after the commit. So, if you can have commit be the last thing in the try, the catch can remain simple. You really need a catch(EntityNotFoundException) that does debug logging and rollback, and a general catch(Exception) that does error logging and rollback.

I also don't see what checking for a DeletedArtifactEvent does here. if you get an EntityNotFoundException from the lock attempt, it means the local artifact is (now) gone so the message is actually wrong. It's more like "I thought I saw X, but now it's gone so nevermind" :-)

And the finally block to detect open transactions!! Technically, the rollback calls should be in little try/catch of their own (the compiler lets you get away with it because spring framework uses RuntimeException for everything) because they can fail as well and it would be good to log.error such a fail. I might have to put a thrown checked exception in TransactionManager...

aside: if you have rollback in the try block, that's usually a sign that you are doing something wrong.; it's also usually a bad sign when the startTransaction() and the commitTransaction() are not in the same block (eg here they are at different indent levels with conditional, which is always dangerous)...  I wouldn't say always, but usually a bad sign.

Reply:
Reworked the transactions and error handling.